pkg/media/h264reader/h264reader.go

```diff
@@ -0,0 +1,272 @@
+// parser code written by https://github.com/chertov
+
+package h264reader
+
+import (
+	"fmt"
+	"io"
+	"log"
+	"strconv"
+)
+
+// H264Reader reads data from stream and constructs h264 nal units
+type H264Reader struct {
+	stream io.Reader
+}
+
+// NewReader creates new H264Reader
+func NewReader(in io.Reader) (*H264Reader, error) {
+	if in == nil {
+		return nil, fmt.Errorf("stream is nil")
+	}
+
+	reader := &H264Reader{
+		stream: in,
+	}
+
+	return reader, nil
+}
+
+// NAL H.264 Network Abstraction Layer
+type NAL struct {
+	PictureOrderCount uint32
+
+	// NAL header
+	ForbiddenZeroBit bool
+	RefIdc           uint8
+	UnitType         nalUnitType
+
+	Data []byte // header byte + rbsp
+}
+
+// ReadFrames reads all data from stream and returns array of all parsed nal units
+func (reader *H264Reader) ReadFrames() []NAL {
+
+	nalsBytes := make([][]byte, 0)
+	nalStream := newFindNalState()
+	for {
+		buf := make([]byte, 1024)
+		n, err := reader.stream.Read(buf)
+		if err != nil && err != io.EOF {
+			log.Fatal("Error Reading: ", err)
+			break
+		}
+		if n == 0 {
+			break
+		}
+		nal := nalStream.NalScan(buf[0:n])
+		nalsBytes = append(nalsBytes, nal...)
+	}
+
+	i := 0
+	var nals []NAL
+	for {
+		if i >= len(nalsBytes) {
+			break
+		}
+		nalData := nalsBytes[i]
+		i = i + 1
+		nal := newNal()
+		nal.parseHeader(nalData[0])
+		if nal.UnitType == sei {
+			continue
+		}
+		nal.Data = nalData
+		nals = append(nals, nal)
+	}
+	return nals
+}
+
+// Implementation
+
+type findNalState struct {
+	PrefixCount   int
+	LastNullCount int
+	buf           []byte
+}
+
+func newFindNalState() findNalState {
+	return findNalState{PrefixCount: 0, LastNullCount: 0, buf: make([]byte, 0)}
+}
+
+func (h *findNalState) NalScan(data []byte) [][]byte {
+	if len(h.buf) > 1024*1024 {
+		panic("findNalState buf len panic")
```

Comment:
We don't want do `panic` inside the library. Instead return `[][]byte, error` but since we need to do a major refactor this might not mater!

Reply:
ok